File: A2/hidato.py
import os
import sys
import time
import subprocess
from subprocess import Popen, PIPE

if len(sys.argv) < 2:
    sys.exit('Usage: %s <problem size>' % sys.argv[0])
    
# Output file
outputFormulaFileName = "hidato-formula.smt2"
outputFormulaFile = open(outputFormulaFileName, 'w')

map = dict()
inputs = []

neighbors = [
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1], [0, 1],
    [1, -1], [1, 0], [1, 1]
]

inputFile = open(sys.argv[1], 'r')

for line in inputFile.readlines():
    inputs.append(line.replace("\n","").split(" "))

rows = len(inputs);
cols = len(inputs[0]);
    
def buildVarName(row, col):
    return "m" + str(row) + str(col)
    
# Declare variables
def declareVar():
    outputFormulaFile.write(";; Declaring variables\n")
    for row in range(0, rows):
        for col in range(0, cols):
            if (inputs[row][col] == "*"):
                continue
            map[buildVarName(row, col)] = True;
            outputFormulaFile.write("(declare-const " + buildVarName(row, col) + " Int)\n")
            if (inputs[row][col] != "-"):
                outputFormulaFile.write("(assert (= " + buildVarName(row, col) + " " + inputs[row][col] + "))\n")
                
            
# Ensures all values are greater than 0
def allGreaterThan0():
    outputFormulaFile.write(";; Ensures all cell are greater than 0\n")
    str = "(assert (and "
    for row in range(0, rows):
        for col in range(0, cols):
            if (inputs[row][col] == "*"):
                continue
            str += "(> " + buildVarName(row, col) + " 0) "
    str += "))\n"
    outputFormulaFile.write(str)

# Ensure there exist some value bigger & smaller than current value by one in the neihborhoods
def checkAllNeighbors():
    outputFormulaFile.write(";; Ensure there exist some value bigger/smaller than current value by one in the neihborhoods\n")
    for row in range(0, rows):
        for col in range(0, cols):
            if (inputs[row][col] == "*"):
                continue
            checkNeighbors(row, col)

# Ensure there exist some value bigger & smaller than current value by one in the neihborhoods
def checkNeighbors(x, y):
    strPlus = "(assert (or "
    strMinus = "(assert (or "
    strBiggest = "(and "
    strSmallest = "(and "
    
    for coordinate in neighbors:
        newX = x + coordinate[0]
        newY = y + coordinate[1]
        if (newX >= 0 and newX < rows and newY >= 0 and newY < cols):
            if (inputs[newX][newY] != "*"):
                current = buildVarName(x,y)
                neighbor = buildVarName(newX,newY)
                strPlus += "(and (<= " + neighbor + " (+ " + current + " 1)) (>= " + neighbor + " (+ " + current + " 1)))"
                strMinus += "(and (<= " + neighbor + " (- " + current + " 1)) (>= " + neighbor + " (- " + current + " 1)))"
                if (inputs[x][y] != "-"):
                    # Ensures that the number could be the smallest or the biggest value in the matrix
                    strBiggest += "(> " + current + " " + neighbor + ")"
                    strSmallest += "(< " + current + " " + neighbor + ")"
        
    strBiggest += ")"
    strSmallest += ")"
    
    if strBiggest == "(and )":
        strBiggest =""
       
    if strSmallest == "(and )":
        strSmallest =""
       
    if strPlus != "(assert (or ":
        strPlus +=  strBiggest + "))"
        outputFormulaFile.write(strPlus + "\n")
    
    if strMinus != "(assert (or ":
        strMinus += strSmallest + "))"
        outputFormulaFile.write(strMinus + "\n")

# Ensures that there is no duplicate value
def noDuplicate():
    outputFormulaFile.write(";; Ensures that there is no duplicate values\n")
    track = dict()
    for row in range(0, rows):
        for col in range(0, cols):
            if (inputs[row][col] == "*"):
                continue
            isEmpty = True
            str = "(assert (and "
            for name in map:
                if (name != buildVarName(row, col) and not track.has_key(name)):
                    str += "(not (= " + buildVarName(row, col) + " " + name  + "))"
                    isEmpty = False
            str += "))\n";
            if (isEmpty == False):
                outputFormulaFile.write(str)
            track[buildVarName(row, col)] = True

def formulateZ3Code():
    declareVar()
    allGreaterThan0()
    checkAllNeighbors()
    noDuplicate()
    outputFormulaFile.write("(check-sat)\n")
    outputFormulaFile.write("(get-model)\n")
    outputFormulaFile.close()

def executeZ3Code(z3Result):
    # Execute the z3 code and fetch the result
    list = dict()
    position = None
    value = None
    outputMatrixFileName = "hidato-output.txt"
    outputMatrixFile = open(outputMatrixFileName, 'w')
    z3ResultLines = z3Result.split("\n")
    
    # Check if there is no solution
    if "unsat" in z3ResultLines[0]:
        outputMatrixFile.write("NO SOLUTION")
        return

    # Parse the output in matrix form
    for line in z3ResultLines:
        if "(define-fun" in line:
            position = line.split("(define-fun ")[1].split(" () Int")[0]
        elif ")" in line:
            value = line.split(")")[0].strip()
        if position != None and value != None:
            list[position] = value
            position = None
            value = None
        
    for row in range(0, rows):
        line = ""
        for col in range(0, cols):
            str1 = "m" + str(row) + str(col)
            if list.has_key(str1):
                line += list[str1] + " "
            else:
                line += "* "
        outputMatrixFile.write(line + "\n")

    outputMatrixFile.close()

formulateZ3Code()
#z3ExecuablePath = './z3/bin/z3.exe'
z3ExecuablePath = '/u/csc410h/fall/pub/z3/bin/z3'
start_time = time.time()
process = Popen([z3ExecuablePath, outputFormulaFileName], stdout=PIPE, stderr=PIPE)
z3Result, stderr = process.communicate()
#print("--- %s seconds ---" % (time.time() - start_time))
#print z3Result
executeZ3Code(z3Result)

File: A2/hidato.js
/*
*  *  7  6  -  *  *
*  *  5  -  -  *  *
31 -  -  4  -  -  18
-  33 2  12 15 19 -
29 28 1  14 -  -  -
*  *  -  24 22 *  *
*  *  25 -  -  *  *

*  *  7  6  9  *  *
*  *  5  8  10 *  *
31 32 3  4  11 16 18
30 33 2  12 15 19 17
29 28 1  14 13 21 20
*  *  27 24 22 *  *
*  *  25 26 23 *  *

*  *  1  2 
*  *  3  - 
8  -  -  4 

*  *  1  2 
*  *  3  5 
8  7  6  4 
*/
var input = [
    ["*","*","7","6","-","*","*"],
    ["*","*","5","-","-","*","*"],
    ["31","-","-","4","-","-","18"],
    ["-","33","2","12","15","19","-"],
    ["29","28","1","14","-","-","-"],
    ["*","*","-","24","22","*","*"],
    ["*","*","25","-","-","*","*"]
]

var test_input = [
    ["*","*","1","2"], 
    ["*","*","3","-"], 
    ["8","-","-","4"]
];

var test_input2 = [
    ["*","*","1","-","-","*","*"],
    ["*","*","-","-","-","*","*"],
    ["*","*","-","-","-","*","*"],
    ["*","*","-","-","-","*","*"],
    ["*","17","-","-","-","-","*"]
]

var map = [];
var neighbors = [
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1], [0, 1],
    [1, -1], [1, 0], [1, 1]
];

// Declare variables
function declareVar(input) {
    var rows = input.length;
    var cols = input[0].length;
    for (var row = 0; row < rows; row++) {
        for (var col = 0; col < cols; col++) {
            if (input[row][col] === "*") {
                continue;
            }
            map[varName(row, col)] = true;
            console.log("(declare-const " + varName(row, col) + " Int)");

            if (input[row][col] !== "-") {
                console.log("(assert (= " + varName(row, col) + " " + input[row][col] + "))");
            }
        }
    }
}

function varName(row, col) {
    return "m" + row + col;
}

// Ensures all values are greater than 0
function allGreaterThan0(input) {
    var str = "(assert (and ";
    var rows = input.length;
    var cols = input[0].length;
    for (var row = 0; row < rows; row++) {
        for (var col = 0; col < cols; col++) {
            if (input[row][col] === "*") {
                continue;
            }
            str += "(> " + varName(row, col) + " 0) ";
        } 
    }
    str += "))";
    console.log(str);
}

function checkAllNeighbors(input) {
    var rows = input.length;
    var cols = input[0].length;
    for (var row = 0; row < rows; row++) {
        for (var col = 0; col < cols; col++) {
            if (input[row][col] === "*") {
                continue;
            }
            checkNeighbors(input, row, col);
        } 
    }
}

// Ensure there exist some value bigger/smaller than current value by one in the neihborhoods
function checkNeighbors(input, x, y) {
    if (input[x][y] !== "-") {
        return;
    }
    var strPlus = "(assert (or ";
    var strMinus = "(assert (or ";
    var rows = input.length;
    var cols = input[0].length;
    for (var i in neighbors) {
        var newX = x + neighbors[i][0];
        var newY = y + neighbors[i][1];
        if (newX >= 0 && newX < rows && newY >= 0 && newY < cols) {
            if (input[newX][newY] !== "*") {
                var current = varName(x,y);
                var neighbor = varName(newX,newY);
                strPlus += "(and (<= " + neighbor + " (+ " + current + " 1)) (>= " + neighbor + " (+ " + current + " 1)))"; 
                strMinus += "(and (<= " + neighbor + " (- " + current + " 1)) (>= " + neighbor + " (- " + current + " 1)))"; 
            } 
        }
    }
    strPlus +="))";
    strMinus +="))";
    console.log(strPlus);
    console.log(strMinus);
}

// Ensures that there is no duplicate value
function noDuplicate(input) {
    var rows = input.length;
    var cols = input[0].length;
    var track = [];
    for (var row = 0; row < rows; row++) {
        for (var col = 0; col < cols; col++) {
            if (input[row][col] === "*") {
                continue;
            }
            var isEmpty = true;
            var str = "(assert (and ";
            for (var name in map) {
                if (name !== varName(row, col) && track[name] === undefined) {
                    str += "(not (= " + varName(row, col) + " " + name  + "))";
                    isEmpty = false;
                }
            }
            str += "))";
            if (isEmpty === false) {
                console.log(str);
            }
            track[varName(row, col)] = true;
        }
    }
}

function parse(input) {
    declareVar(input);
    allGreaterThan0(input);
    checkAllNeighbors(input);
    noDuplicate(input);
    console.log("(check-sat)");
    console.log("(get-model)");
}


parse(input);


/*
delete
m42 smallest
m31 largest

(< m52 m42)
(= (m42 (+ m52 1)))
(> m30 m31)
(= (m31 (- m30 1)))
(> m21 m31)
(= (m31 (- m21 1)))
(= (m31 (- m22 1)))
(> m22 m31)
*/